src/ui/public/index_patterns/_index_pattern.js

```diff
@@ -13,318 +13,334 @@ import IndexPatternsFieldListProvider from 'ui/index_patterns/_field_list';
+
+          return safeConfirm(confirmMessage)
+            .then(() => Promise
+              .try(() => {
+                const cached = patternCache.get(this.id);
                 if (cached) {
                   return cached.then(pattern => pattern.destroy());
                 }
               })
-              .then(() => docSource.doIndex(body))
-              .then(setId);
-            },
-            _.constant(false) // if the user doesn't overwrite, resolve with false
-          );
-        }
-        return Promise.resolve(false);
-      });
-    };
-
-    self.save = function () {
-      let body = self.prepBody();
-      return docSource.doIndex(body).then(setId);
-    };
-
-    self.refreshFields = function () {
-      return mapper.clearCache(self)
-      .then(self._fetchFields)
-      .then(self.save);
-    };
-
-    self._fetchFields = function () {
-      return mapper.getFieldsForIndexPattern(self, true)
-      .then(function (fields) {
-        // append existing scripted fields
-        fields = fields.concat(self.getScriptedFields());
-
-        // initialize self.field with this field list
-        initFields(fields);
-      });
-    };
+              .then(() => this._docSource.doIndex(body))
+              .then(id => this._setId(id)),
+              _.constant(false) // if the user doesn't overwrite, resolve with false
+            );
+        });
+    }
 
-    self.toJSON = function () {
-      return self.id;
-    };
+    save() {
+      const body = this.prepBody();
+      return this._docSource
+        .doIndex(body)
+        .then(id => this._setId(id));
+    }
 
-    self.toString = function () {
-      return '' + self.toJSON();
-    };
+    refreshFields() {
+      return mapper
+        .clearCache(this)
+        .then(() => this._fetchFields())
+        .then(() => this.save());
+    }
 
-    self.destroy = function () {
-      patternCache.clear(self.id);
-      docSource.destroy();
-    };
+    _initFields(fields) {
+      this._fieldsInitialized = true;
+      this.fields = new FieldList(this, fields || this.fields || []);
+    }
 
-    self.metaFields = config.get('metaFields');
-    self.getComputedFields = getComputedFields.bind(self);
+    _fetchFields() {
+      return mapper
+        .getFieldsForIndexPattern(this, true)
+        .then(fields => {
+          const scripted = this.getScriptedFields();
+          const all = fields.concat(scripted);
+          this._initFields(all);
+        });
+    }
 
-    self.flattenHit = flattenHit(self);
-    self.formatHit = formatHit(self, fieldformats.getDefaultInstance('string'));
-    self.formatField = self.formatHit.formatField;
+    toJSON() {
+      return this.id;
+    }
+
+    toString() {
+      return '' + this.toJSON();
+    }
+
+    destroy() {
+      patternCache.clear(this.id);
+      this._docSource.destroy();
+    }
   }
 
-  IndexPattern.prototype.routes = {
+  IndexPattern.routes = {
```

Comment:
Maybe making `routes` a getter on the class would be more equivalent to the old version?